### ISSUE TITLE
Clang differences when using commandline options or pragmas to turn on fast-math functionallity

### ISSUE THREAD

**Issue by karka228** [llvm/llvm-project#87758]
When looking at the output from clang -cc1 for the program below

```
float powf(float x, float y);
float sqrtf(float x);

float f2(float x, int y) {
  x = powf(x,y);
  x = sqrtf(x);
  return x;
}
```
Compiled with:
`$ clang -cc1 foo.c -triple x86_64-linux-gnu -emit-llvm -o - -ffast-math -ffp-contract=fast`

The result of the `-ffast-math` and ` -ffp-contract=fast` options is that both calls to pow and sqrt are annotated with fast math flag `fast` (see LangRef):

```
  %2 = call fast float @llvm.pow.f32(float %0, float %conv)
  %4 = call fast float @llvm.sqrt.f32(float %3)
```

When instead turning on fast math with `#pragma float_control(precise,off)` like this:

```
float powf(float x, float y);
float sqrtf(float x);

#pragma float_control(push)
#pragma float_control(precise,off)

float f2(float x, int y) {
  x = powf(x,y);
  x = sqrtf(x);
  return x;
}
#pragma float_control(pop)
```
Compiled with:
`$ clang -cc1 foo.c -triple x86_64-linux-gnu -emit-llvm -o -`

The result of the pragma is that only the sqrt call get to be annotated with the fast math flag `fast`:

```
  %2 = call float @llvm.pow.f32(float %0, float %conv)
  %4 = call fast float @llvm.sqrt.f32(float %3)
```

Why are not `llvm.pow.f32` not annotated with fast math flag `fast` when using `#pragma float_control(precise,off)`?

**Comment by dtcxzyw** [comment]
cc @andykaylor @arsenm

**Comment by arsenm** [comment]
I'd guess there's a bug similar to the one fixed by f6f625f4c6c3cbeb8cf6b889cdafc848dd4cb117

**Comment by andykaylor** [comment]
Actually, I think the problem is similar to this: 63f0833ec4beb7d6bc63759ad8b336ad6dce0f93

I noticed that the fast-math flags are applied for calls with one argument, but not for calls with two or three arguments. We have different functions in CGBuiltin.cpp to handle binary and ternary calls. We should probably refactor to unify those in some way to avoid this kind of problem.

cc: @zahiraam
